cpp/tests/utility/vector_equality.hpp

```diff
@@ -16,24 +16,77 @@
 
 #pragma once
 
+#include <cuspatial/vec_2d.hpp>
+
+#include <rmm/device_uvector.hpp>
+#include <rmm/device_vector.hpp>
+
+#include <thrust/host_vector.h>
+
 #include <gmock/gmock.h>
 #include <gtest/gtest.h>
 
 #include <type_traits>
 
+template <typename T>
+std::ostream& operator<<(std::ostream& os, cuspatial::vec_2d<T> const& vec)
+{
+  return os << "(" << vec.x << "," << vec.y << ")";
+}
+
 namespace cuspatial {
 namespace test {
 
+template <typename T>
+auto float_matcher(T val)
+{
+  if constexpr (std::is_same_v<T, float>) {
+    return ::testing::FloatEq(val);
+  } else {
+    return ::testing::DoubleEq(val);
+  }
+}
+
+MATCHER(vec_2d_matcher,
+        std::string(negation ? "are not" : "are") + " approximately equal vec_2d structs")
+{
+  auto lhs = std::get<0>(arg);
+  auto rhs = std::get<1>(arg);
+
+  if (::testing::Matches(float_matcher(rhs.x))(lhs.x) &&
+      ::testing::Matches(float_matcher(rhs.y))(lhs.y))
+    return true;
+
+  //*result_listener << lhs << " != " << rhs;
+
+  return false;
+}
+
+template <typename T>
+constexpr bool is_vec_2d()
```

Comment:
and this in `traits.hpp`?

Reply:
My problem was that traits.hpp stuff is all in the `detail` namespace, and I don't like reaching into `detail` from tests.

Reply:
Moved traits out of detail and added this to traits.hpp.